crypto/dilithium/p_dilithium3_asn1.c

```diff
@@ -47,12 +67,16 @@ static int dilithium3_set_pub_raw(EVP_PKEY *pkey, const uint8_t *in, size_t len)
 
   DILITHIUM3_KEY *key = OPENSSL_malloc(sizeof(DILITHIUM3_KEY));
   if (key == NULL) {
-    OPENSSL_PUT_ERROR(EVP, ERR_R_MALLOC_FAILURE);
+    return 0;
+  }
+  key->pub = OPENSSL_malloc(DILITHIUM3_PUBLIC_KEY_BYTES);
+  if (key->pub == NULL) {
+    OPENSSL_free(key);
     return 0;
   }
 
   OPENSSL_memcpy(key->pub, in, len);
-  key->has_private = 0;
+  key->priv = NULL;
 
   dilithium3_free(pkey);
   pkey->pkey.ptr = key;
```

Comment:
before assigning `key` to `pkey->pkey.ptr ` should we free `pkey->pkey.ptr ` if it's already pointing to an allocated key? same for the `_set_priv` function.

Reply:
i checked what other `_set_pub` and `_set_pub` functions do. [x25519 and ed25519 appear to be](https://github.com/search?q=repo%3Aaws%2Faws-lc+_set_priv&type=code) the only other implementors of these funcitons. neither free `pkey->pkey.ptr`'s content prior to overwriting it. i suppose that we could break the pattern here, but i worry (slightly) about some other piece of code retaining a reference to `ptr`'s contents and causing a use-after-free somewhere else in the codebase...

Reply:
great, thanks for checking!